jail/20230308.md

```diff
@@ -0,0 +1,176 @@
+**Dave**: Every jail is based on a commit hash... check the hash.
+
+Note the possibility of [integer mix-ups here](https://github.com/mekanix/freebsd-src/blob/feature/jail/usr.sbin/jail/config.c#L204-L206). This will give Jail ID "something" "3":
+
+```
+* { $id = 1; }
+something { $id = 2; }
+* { $id = 3; }
+```
+
+Anything global needs to be last, not first. Feature? Bug?
+
+Note `/usr/sbin/ctlduse` a local configuration file for a UCL example.
+
+Is there an in-base UCL parser equivalent to `jq(1)`? `libucl` is there.
+
+Is it helpful to inject runtime options that are not persistent (in a configuration file)?
+
+You _can_ [obtain the runtime configuration from a running jail](https://dan.langille.org/2019/10/23/moving-poudriere-from-the-host-into-a-jail/). Not obvious in the manual  
+page: Try this command:`jls -n -j JID`
+
+`jls -vh -j app0` will give values, `-h` for headers.
+
+The order of flags matters!`-vh` and `-hv` do different things!
+
+**Question** : Is the `-h` header output consistent?
+
+`jls -d` will show deceased jails and will give a history. Some Jails never truly die! Is this a ZFS issue where resources are not released? There is a GOTO loop that monitors the Jail and ZFS subsystems. This might be only a 14-CURRENT issue. Issue went away on UFS. Saw the  
+same issue on VNET. Needed -vnet flag. This was found with [OccamBSD](https://github.com/michaeldexter/occambsd).
+
+Dan's patch (link?): Prefer not specify running Jail names and prefer have global variables in  
+`jail.conf`, rather than repeating information. Note this [Ansible playbook](https://git.langille.org/dvl/ansible/src/branch/main/jail.conf.d) for some of this.
+
+**Antranig** : Note the [`sysctl.conf.d`](https://reviews.freebsd.org/D32128) review.
+
+Note `cron.d` that is often used by packages in `/usr/local/etc`... Note `syslog.d` linked from `syslog.conf`, and the "new syslog" syntax. The syntax changed in 13!
+
+**Question** : How to stagger Jail operations in cron? Dan: Use jitter?
+
+**Poll** : Who is using DHCP with Jails? Dan is, Goran runs it in Jails with IPv4/IPv6. Run rtdvd/rtsol on the router... (advertisements/solicitation) Expiration(s) in 10 minutes. Can result in two default routes. The rtsol route does not disappear.
+
+**Question:** Should FreeBSD Jails be OCI-compliant and follow their standard? Too many implementations become standards. To what degree is runj following the standard? Note Bryan Cantrill's [talk](https://www.youtube.com/watch?v=AdMqCUhvRz8) on debugging Docker in Production. Consider a Linux-compatibility shim? The Go-based control plane is quite decoupled. Nomad scales much better - produced a 2000 node cluster. [Mesos](https://mesos.apache.org/) scales much better than Kubernetes or Nomad
+[Link](https://mesosphere.github.io/marathon/)
+
+**Note:** Brendan Gregg's comments on tracing Zones and Jails compared to Linux containers.
+
+**Quote:** PHK: "I used a script to create lots of jails and got bored at 64k"
```

Comment:
I don't know where that quote is from nor why was it pasted. Given the context it is supposed to justify a claim there is no jail-management performance problem?

According to git log jails showed up in 1999 and they were incredibly primitive in the beginning.

```
        MALLOC(pr, struct prison *, sizeof *pr , M_PRISON, M_WAITOK);
        bzero((caddr_t)pr, sizeof *pr);
        error = copyinstr(j.hostname, &pr->pr_host, sizeof pr->pr_host, 0);
        if (error) 
                goto bail;
        pr->pr_ip = j.ip_number;

        ca.path = j.path;
        error = chroot(p, &ca);
        if (error)
                goto bail;

        pr->pr_ref++;
        p->p_prison = pr;
        p->p_flag |= P_JAILED;
```

et voila, new jail in place.

One thing you may notice there is no ID associated with it and it did not land in any data structure allowing to find it later.

There is 0 surprise something like this managed to get 64k instances and not be affected.

In contrast, after jail_attach was implemented in 2003, IDs also showed up and a global *linked list* was added. There is *linear scans* needed to jail_attach, does not look good for 64k jails, does it.

Moreover, the current tolling calls jexec several times to de facto spawn a jail, making the situation worse.

Bare minimum the linear scan needs to be whacked, the easiest way out which may also happen to be tolerable is a RB tree over it. Even then note finding a jail is a globally serializing operation when it does not have to be. I did mention this to jamie@.

That aside I have to note both nested and unprivileged jails are nasty cans of worms.